File: ipv6_logging.py
#!/usr/bin/python
from bcc import BPF

import sys
import socket
import os

from sys import argv

#arguments
interface="eth0"

if len(argv) == 3:
    if str(argv[1]) == '-i':
        interface = argv[2]

print("USAGE: %s [-i <if_name>]" % argv[0])
print ("binding socket to '%s'" % interface)	
 

bpf = BPF(src_file = "ipv6_filter.c", debug = 0)

function_tcp_filter = bpf.load_func("ipv6_filter", BPF.SOCKET_FILTER)

BPF.attach_raw_socket(function_tcp_filter, interface)

socket_fd = function_tcp_filter.sock

sock = socket.fromfd(socket_fd, socket.PF_PACKET, socket.SOCK_RAW, socket.IPPROTO_IP)

sock.setblocking(True)

ETH_HLEN = 14 #Ethernet header length in bytes
IPV6_ADDR_SHIFT = 8 #Shift of addresses in IPv6 header

while 1:
    packet_str = os.read(socket_fd, 2048)
    
    packet_bytearray = bytearray(packet_str)

    src_host_ip = packet_bytearray[ETH_HLEN + IPV6_ADDR_SHIFT : ETH_HLEN + IPV6_ADDR_SHIFT + 16] 
    dst_host_ip = packet_bytearray[ETH_HLEN + IPV6_ADDR_SHIFT + 16: ETH_HLEN + IPV6_ADDR_SHIFT + 32]

    src_host_ip = []
    dst_host_ip = []

    for i in range(ETH_HLEN + IPV6_ADDR_SHIFT, ETH_HLEN + IPV6_ADDR_SHIFT + 16, 2):
	src_host_ip.append(packet_bytearray[i] * 256 + packet_bytearray[i + 1])

    for i in range(ETH_HLEN + IPV6_ADDR_SHIFT + 16, ETH_HLEN + IPV6_ADDR_SHIFT + 32, 2):
        dst_host_ip.append(packet_bytearray[i] * 256 + packet_bytearray[i + 1])

    src_str = ':'.join(hex(x)[2:] for x in src_host_ip)
    dst_str = ':'.join(hex(x)[2:] for x in dst_host_ip)

    msg = "Src host: " + src_str + "   Dest host: "+ dst_str

    print (msg)

File: ipv6_logging.cc
#include <unistd.h>
#include <fstream>
#include <iostream>
#include <string>
#include <sys/socket.h>
#include <linux/ipv6.h>
#include <linux/kernel.h>

#include "bcc/BPF.h"

const std::string BPF_PROGRAM = R"(
#include <uapi/linux/ptrace.h>
#include <net/sock.h>
#include <bcc/proto.h>

#define IPv6 0x86DD

int ipv6_filter(struct __sk_buff *skb) { 
	u8 *cursor = 0;	

	struct ethernet_t *ethernet = cursor_advance(cursor, sizeof(*ethernet));
	
	if(ethernet->type != IPv6) {
	    	goto DROP;
	}

	KEEP:
		return -1;

	DROP:
		return 0;
})";

const int ETH_HLEN = 14;
const int IPV6_ADDR_SHIFT = 8;
const int IPV6_ADDR_LEN = 16;

void printIPv6Address(unsigned char* buf, int start_address_byte) {
    for (int i = start_address_byte; i < start_address_byte + IPV6_ADDR_LEN; i += 2) {
        std::cout << (int)buf[i] * 256 + (int)buf[i + 1];
        if (i + 2 < start_address_byte + IPV6_ADDR_LEN) {
            std::cout << ":";
        }
    }
}

int main(int argc, char* argv[]) {

    std::cout << "USAGE: "<< argv[0] << " [-i <if_name>]" << std::endl;

    std::string interface = "eth0";

    if (argc == 3 && strcmp(argv[1], "-i") == 0) {
        interface = argv[2];
    }

    std::cout << "interface = " << interface << std::endl;

    ebpf::BPF bpf;
    auto init_res = bpf.init(BPF_PROGRAM);

    if (init_res.code() != 0) {
        std::cerr << init_res.msg() << std::endl;
        return 1;
    }

    int prog_fd;
    auto load_res = bpf.load_func("ipv6_filter", BPF_PROG_TYPE_SOCKET_FILTER, prog_fd);

    if (load_res.code() != 0) {
        std::cerr << load_res.msg() << std::endl;
        return 1;
    }

    int sock = bpf_open_raw_sock(interface.c_str());

    if (bpf_attach_socket(sock, prog_fd) != 0){
        std::cerr << "Error creating socket" << std::endl;
        return 1;
    }

    std::cout << "Socket attached" << std::endl;

    unsigned char *buf = new unsigned char[2048];
    long size;
    while (true) {
        size = read(sock, buf, 2048);
        if(size != -1) {
            std::cout << std::hex << "Src host: ";
            printIPv6Address(buf, ETH_HLEN + IPV6_ADDR_SHIFT);

            std::cout << "     Dest host: ";
            printIPv6Address(buf, ETH_HLEN + IPV6_ADDR_SHIFT + IPV6_ADDR_LEN);

            std::cout << std::endl;
        }
    }
}
